p2p/libp2p/discovery/optimizedKadDhtDiscoverer.go

```diff
@@ -0,0 +1,231 @@
+		return nil, p2p.ErrInvalidSeedersReconnectionInterval
+	}
+
+	okdd := &optimizedKadDhtDiscoverer{
+		sharder:                     sharder,
+		peersRefreshInterval:        arg.PeersRefreshInterval,
+		seedersReconnectionInterval: arg.SeedersReconnectionInterval,
+		protocolID:                  arg.ProtocolID,
+		initialPeersList:            arg.InitialPeersList,
+		bucketSize:                  arg.BucketSize,
+		routingTableRefresh:         arg.RoutingTableRefresh,
+		status:                      statNotInitialized,
+		chanInit:                    make(chan struct{}),
+		errChanInit:                 make(chan error),
+		chanConnectToSeeders:        make(chan struct{}),
+		chanDoneConnectToSeeders:    make(chan struct{}),
+	}
+
+	okdd.initKadDhtHandler = okdd.initKadDht
+	okdd.hostConnManagement, err = NewHostWithConnectionManagement(arg.Host, okdd.sharder)
+	if err != nil {
+		return nil, err
+	}
+
+	go okdd.processLoop(arg.Context)
+
+	return okdd, nil
+}
+
+// Bootstrap will start the bootstrapping new peers process
+func (okdd *optimizedKadDhtDiscoverer) Bootstrap() error {
+	okdd.chanInit <- struct{}{}
+	return <-okdd.errChanInit
+}
+
+func (okdd *optimizedKadDhtDiscoverer) processLoop(ctx context.Context) {
+	chTimeSeedersReconnect := time.After(okdd.seedersReconnectionInterval)
+	chTimeFindPeers := time.After(okdd.peersRefreshInterval)
+
+	for {
+		select {
+		case <-okdd.chanInit:
+			err := okdd.init(ctx)
+			okdd.errChanInit <- err
+			okdd.connectToSeeders(ctx, false)
+			okdd.findPeers(ctx)
+
+		case <-chTimeSeedersReconnect:
+			isConnectedToSeeders := okdd.connectToSeeders(ctx, false)
+			chTimeSeedersReconnect = okdd.createChTimeSeedersReconnect(isConnectedToSeeders)
+
+		case <-okdd.chanConnectToSeeders:
+			isConnectedToSeeders := okdd.connectToSeeders(ctx, true)
+			chTimeSeedersReconnect = okdd.createChTimeSeedersReconnect(isConnectedToSeeders)
+			okdd.chanDoneConnectToSeeders <- struct{}{}
+
+		case <-chTimeFindPeers:
+			okdd.findPeers(ctx)
+			chTimeFindPeers = time.After(okdd.peersRefreshInterval)
+
+		case <-ctx.Done():
+			log.Debug("closing the p2p bootstrapping process")
+			return
+		}
+	}
+}
+
+func (okdd *optimizedKadDhtDiscoverer) createChTimeSeedersReconnect(isConnectedToSeeders bool) <-chan time.Time {
+	if isConnectedToSeeders {
+		//the reconnection will be done less often
+		return time.After(okdd.seedersReconnectionInterval)
+	}
+
+	// no connection to seeders, let's try a little bit faster
+	return time.After(okdd.peersRefreshInterval)
+}
+
+func (okdd *optimizedKadDhtDiscoverer) init(ctx context.Context) error {
+	if okdd.status != statNotInitialized {
+		return p2p.ErrPeerDiscoveryProcessAlreadyStarted
+	}
+
+	kadDhtHandler, err := okdd.initKadDhtHandler(ctx)
+	if err != nil {
+		return err
+	}
+
+	okdd.kadDHT = kadDhtHandler
+	okdd.status = statInitialized
+
+	return nil
+}
+
+func (okdd *optimizedKadDhtDiscoverer) initKadDht(ctx context.Context) (KadDhtHandler, error) {
```

Comment:
This is more like a create instead of init?

Reply:
can be renamed

Reply:
done